sdks/java/extensions/zetasketch/src/main/java/org/apache/beam/sdk/extensions/zetasketch/HllCountInitFn.java

```diff
@@ -0,0 +1,152 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.zetasketch;
+
+import static org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Preconditions.checkArgument;
+
+import com.google.zetasketch.HyperLogLogPlusPlus;
+import com.google.zetasketch.shaded.com.google.protobuf.ByteString;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.CoderRegistry;
+import org.apache.beam.sdk.transforms.Combine;
+
+/**
+ * {@link Combine.CombineFn} for the {@link HllCount.Init} combiner.
+ *
+ * @param <InputT> type of input values to the function
+ * @param <HllT> type of the HLL++ sketch to compute
+ */
+abstract class HllCountInitFn<InputT, HllT>
+    extends Combine.CombineFn<InputT, HyperLogLogPlusPlus<HllT>, byte[]> {
+
+  private int precision;
+
+  private HllCountInitFn() {
+    setPrecision(HllCount.DEFAULT_PRECISION);
+  }
+
+  int getPrecision() {
+    return precision;
+  }
+
+  void setPrecision(int precision) {
+    checkArgument(
+        precision >= HllCount.MINIMUM_PRECISION && precision <= HllCount.MAXIMUM_PRECISION,
+        "Invalid precision: %s. Valid range is [%s, %s].",
+        precision,
+        HllCount.MINIMUM_PRECISION,
+        HllCount.MAXIMUM_PRECISION);
+    this.precision = precision;
+  }
+
+  @Override
+  public Coder<HyperLogLogPlusPlus<HllT>> getAccumulatorCoder(
+      CoderRegistry registry, Coder<InputT> inputCoder) {
+    return HyperLogLogPlusPlusCoder.of();
+  }
+
+  @Override
+  public HyperLogLogPlusPlus<HllT> mergeAccumulators(
+      Iterable<HyperLogLogPlusPlus<HllT>> accumulators) {
+    HyperLogLogPlusPlus<HllT> merged = createAccumulator();
+    for (HyperLogLogPlusPlus<HllT> accumulator : accumulators) {
+      merged.merge(accumulator);
+    }
+    return merged;
+  }
+
+  @Override
+  public byte[] extractOutput(HyperLogLogPlusPlus<HllT> accumulator) {
+    return accumulator.serializeToByteArray();
+  }
+
+  static HllCountInitFn<Integer, Integer> forInteger() {
+    return new ForInteger();
+  }
+
+  static HllCountInitFn<Long, Long> forLong() {
+    return new ForLong();
+  }
+
+  static HllCountInitFn<String, String> forString() {
+    return new ForString();
+  }
+
+  static HllCountInitFn<byte[], ByteString> forBytes() {
+    return new ForBytes();
+  }
+
+  private static class ForInteger extends HllCountInitFn<Integer, Integer> {
+
+    @Override
+    public HyperLogLogPlusPlus<Integer> createAccumulator() {
+      return new HyperLogLogPlusPlus.Builder().normalPrecision(getPrecision()).buildForIntegers();
+    }
+
+    @Override
+    public HyperLogLogPlusPlus<Integer> addInput(
+        HyperLogLogPlusPlus<Integer> accumulator, Integer input) {
+      accumulator.add(input.intValue());
```

Comment:
HyperLogLogPlusPlus has a add(V value) method which does the same (as you've probably seen), but it does include an extra switch statement, so you could simplify the code here, but I do agree that the current version is slightly more efficient! :)

Reply:
Yeah, I have seen that. Regardless of efficiency, I could not simplify the code by implementing the `addInput()` function in the super class. This is because the `bytesSketch` version need to call `add(byte[])` instead of `add(V)`. If we call `add(V)` then inside the function the type check will fail (`byte[]` is not `ByteString`).

Reply:
I guess you could implement the method in the superclass for Long, Int and String, and then override it for byte[]? Having said that, the custom implementations for each type (= current status) also LGTM.

Reply:
Yeah, that's something I could do. But let's keep the current status for the efficiency benefit. :)